crates/sled/src/iter.rs

```diff
@@ -1,48 +1,182 @@
-use std::ops;
+use std::ops::{Bound, Bound::{Included, Excluded, Unbounded}};
+use std::iter::FusedIterator;
 
 use pagecache::{Measure, M};
 
 use super::*;
 
-fn lower_bound_includes<'a>(lb: &ops::Bound<Vec<u8>>, item: &'a [u8]) -> bool {
-    match lb {
-        ops::Bound::Included(ref start) => start.as_slice() <= item,
-        ops::Bound::Excluded(ref start) => start.as_slice() < item,
-        ops::Bound::Unbounded => true,
+fn lower_bound_includes(bound: &Bound<Vec<u8>>, item: &[u8]) -> bool {
+    match bound {
+        Included(start) => start.as_slice() <= item,
+        Excluded(start) => start.as_slice() < item,
+        Unbounded => true,
     }
 }
 
-fn upper_bound_includes<'a>(ub: &ops::Bound<Vec<u8>>, item: &'a [u8]) -> bool {
-    match ub {
-        ops::Bound::Included(ref end) => item <= end.as_ref(),
-        ops::Bound::Excluded(ref end) => item < end.as_ref(),
-        ops::Bound::Unbounded => true,
+fn upper_bound_includes(bound: &Bound<Vec<u8>>, item: &[u8]) -> bool {
+    match bound {
+        Included(end) => item <= end.as_slice(),
+        Excluded(end) => item < end.as_slice(),
+        Unbounded => true,
     }
 }
 
+fn node_of_key<'a>(tree: &Tree, key: &[u8], tx: &'a Tx) -> Result<(PageId, &'a Node)> {
+    let (page_id, frag, _tree_ptr) =
+        tree.path_for_key(key, tx)?
+            .pop()
+            .expect("path should never be empty");
+
+    let node = frag.unwrap_base();
+
+    Ok((page_id, node))
+}
+
+fn node_of_pageid<'a>(tree: &Tree, page_id: PageId, tx: &'a Tx) -> Result<&'a Node> {
+    let res = tree
+        .context
+        .pagecache
+        .get(page_id, tx)
+        .map(|page_get| page_get.unwrap());
+
+    let frag = res.map(|(frag, _ptr)| frag)?;
+    let node = frag.unwrap_base();
+
+    Ok(node)
+}
+
+fn biggest_node<'a>(tree: &Tree, tx: &'a Tx) -> Result<(PageId, &'a Node)> {
+    let (mut page_id, mut node) = node_of_key(tree, b"", tx)?;
+
+    while let Some(next_page_id) = node.next {
+        let next_node = node_of_pageid(tree, next_page_id, tx)?;
+        page_id = next_page_id;
+        node = next_node;
+    }
+
+    Ok((page_id, node))
+}
```

Comment:
I would like to find a better way (e.g. binary search the biggest key) instead of doing this linked list like travel.